### PR TITLE
Implement highlight/hide all for item search results.

### LINKED ISSUE #5148: Item search: User must be able to highlight/hide all item search results.
Currently the user can select an items search result and the selected feature is highlighted. Add a button or a link to the search results view that when clicked highlights all the search results.

This might be easy to do for some data sets like 3d-tiles by constructing a style expression from the search query and applying the style expression to the tileset. For other datasets this might be hard to implement or might involve some brute forcing by iterating all the search results and highlighting it. 

From an API perspective, each catalog item with the `SearchableItemMixin` can implement an optional method called `selectAllResults`. We show the "Highlight all" link in the UI only for catalog items that implement this method.

### PR DESCRIPTION
**Pull request by na9da**
### What this PR does

Fixes #5148 

Implements highlighting/hiding all search results for 3d-tiles. The way it works is by listening for `tileVisible` events for the tileset and tagging each feature with a matching ID and then applying a cesium `color` or `show` expression for all features with that tag.

The alternate solution was to convert a search query to cesium style expression, however this wont work for complex queries involving full text search.

I have checked the performance of the current implementation and it didn't seem too bad.

I had to refactor quite a bit to streamline the whole thing, hence the noisy diff!

### Testing

Using the same dataset from https://github.com/TerriaJS/terriajs/pull/5112#issue-545071441

### Checklist

-   ~[ ] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)~
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

**Review by KeyboardSounds — CHANGES_REQUESTED** [review]
Watching the search feature get better and better is so exciting! "highlight all" in particular adds a lot to how useful it is, making it easier for users to see patterns in their data. Nice work on those one ⭐

IMO, when you hit the search button, by default all of the buildings that match the search criteria should be highlighted, rather than the user having to click "highlight all".

Some issues I spotted when I was testing:
- The size of the text on the HighlightAll buttons seems to change when you click the button. Also, since it's a toggle, it should probably be a checkbox or something rather than a button to make the current state clearer.
- If you click through enough (10+) search results, all of the buildings disappear. This could be a Cesium problem struggling with loading and unloading all the tiles, but it's worth investigating.
- If I select a small (~< 5M) tall building in the search results, it doesn't get highlighted and the feature info panel for it is stuck on loading. Possibly because the tile that contains it never loads?
- If I bring up the search tool and then remove the dataset from the workbench, the search tool stays on screen.

File: lib/Models/Cesium3DTilesCatalogItem.ts [hunk 1]
```diff
@@ -33,26 +39,127 @@ export default class Cesium3DTilesCatalogItem
...
+
+    let pickedFeatures: PickedFeatures | undefined = undefined;
+    if (results.length === 1)
+      pickedFeatures = this.popupInfoPanelForFeature(
+        { [idPropertyName]: results[0].id },
+        SEARCH_RESULT_TAG,
+        60000
+      );
+
+    const watch = (tile: Cesium3DTile) => {
```
[the remark this answers is not in the document]
Comment (na9da) [reply lib/Models/Cesium3DTilesCatalogItem.ts]: Good point! I have extracted it out to a method called `_watchForNewTileFeatures`.

File: lib/ReactViews/Tools/ItemSearchTool/SearchResults.tsx [hunk 2]
```diff
@@ -56,6 +95,15 @@ const SearchResults: React.FC<SearchResultsProps> = props => {
...
       </List>
+      {currentSelection.is === "highlightAll" && (
+        <HighlightResults item={item} results={results} />
+      )}
+      {currentSelection.is === "singleResult" && (
+        <HighlightResults item={item} results={[currentSelection.result]} />
+      )}
+      {currentSelection.is === "hideAll" && (
+        <HideAllResults item={item} results={results} />
+      )}
```
[the remark this answers is not in the document]
Comment (na9da) [reply lib/ReactViews/Tools/ItemSearchTool/SearchResults.tsx]: I have addressed the basic concern of improving clarity by extracting the individual components into a new component called `MapEffects` to make it clear that it is applying some effect to the map.

On the use of renderless components, I think it is a known [pattern](https://shopify-1.gitbook.io/react/4.-advanced-patterns/renderless-components) in react. In this case, like you said, we could bind these effects to the click events but then have to remember to remove the effects when the component unmounts in some other useEffect disposer function. With renderless components we can extract these two related bits of logic into a separate component that takes care of the mount/unmount behaviors. 

In general, I think, renderless components are useful when you want perform some side-effect on mount/update and then cleanly dispose the effect on unmount.

Hope that justifies it :)

**Comment by na9da** [comment]
Thanks for the awesome feedback @KeyboardSounds!

All are great points, I have addressed most of them or provided justifications where necessary.

> IMO, when you hit the search button, by default all of the buildings that match the search criteria should be highlighted, rather than the user having to click "highlight all".

That's a good idea, I have made "highlight all" the default behavior.

> The size of the text on the HighlightAll buttons seems to change when you click the button. Also, since it's a toggle, it should probably be a checkbox or something rather than a button to make the current state clearer.

I have fixed the font-size issue. the checkbox thing is a good suggestion that I'll look at when I work in some of the design changes from @philipgrimmett 's new design - https://github.com/TerriaJS/terriajs/issues/5246

> If you click through enough (10+) search results, all of the buildings disappear. This could be a Cesium problem struggling with loading and unloading all the tiles, but it's worth investigating.

Good catch. This was happening because of a previous change I introduced for calculating modelMatrix from transformation traits.

https://github.com/TerriaJS/terriajs/blob/e0df615a65533aa26d564ba8de9955ebcb0cf8d8/lib/ModelMixins/Cesium3dTilesMixin.ts#L233-L238

It was modifying the modelMatrix and then when `mapItems` recomputes due to some change in styles, it incorrectly dereives new the modelMatrix from the previous modelMatrix and the traits. Whereas the correct behavior is to keep a copy of the original `tileset.root.transform` and re-compute every time from the original transform an the transformation traits. I have fixed the problem below also ensuring that the modelMatrix is re-computed only if the transformation traits change.

https://github.com/TerriaJS/terriajs/blob/20f9f46e52ff30ca897fada241180e488587a5f9/lib/ModelMixins/Cesium3dTilesMixin.ts#L252-L253

> If I select a small (~< 5M) tall building in the search results, it doesn't get highlighted and the feature info panel for it is stuck on loading. Possibly because the tile that contains it never loads?

Yes this is a problem. The only way to fix that would be to provide the lat/lon/height of features [when indexing](https://github.com/TerriaJS/terriajs-indexer/blob/main/README.md#indexer-configuration-options). This helps us zoom in closer to the building. The feature info panel will timeout after 60secs if it cannot resolve the feature.

> If I bring up the search tool and then remove the dataset from the workbench, the search tool stays on screen.

Fixed!

**Comment by na9da** [comment]
Other changes:

Previously we were listening only for `tileVisible` events to apply the highlight. This resulted in a visible flicker because the tile features would render with the original style then the new highlight is applied and there is a delay for the highlight to appear. We now listen for both `tileVisible` and `tileLoad` events, this ensures that the highlight is applied from the start for newly loaded tiles.

Changed the "Hide all" option to "Show only matching" after feedback from Mats & others.

File: lib/ModelMixins/Cesium3dTilesMixin.ts [hunk 3]
```diff
@@ -472,10 +472,36 @@ export default function Cesium3dTilesMixin<
     }
 
     @action
```
[the remark this answers is not in the document]
Comment (KeyboardSounds) [reply lib/ModelMixins/Cesium3dTilesMixin.ts]: @na9da?

File: lib/ReactViews/Tools/ItemSearchTool/SearchResults.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (KeyboardSounds) [reply lib/ReactViews/Tools/ItemSearchTool/SearchResults.tsx]: Consider me convinced! I like the new way you've done it with more of a separation between view and render logic a lot better.

File: lib/ModelMixins/Cesium3dTilesMixin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (na9da) [reply lib/ModelMixins/Cesium3dTilesMixin.ts]: Done!

**Review by KeyboardSounds — APPROVED** [review]

**Pull request merged**